core/src/main/scala/caliban/wrappers/ApolloPersistedQueries.scala

```diff
@@ -1,58 +1,123 @@
 package caliban.wrappers
 
 import caliban.CalibanError.ValidationError
-import caliban.Value.{ NullValue, StringValue }
+import caliban.Value._
+import caliban.execution.ExecutionRequest
 import caliban.parsing.adt.Document
-import caliban.wrappers.Wrapper.{ EffectfulWrapper, OverallWrapper, ParsingWrapper }
+import caliban.wrappers.Wrapper._
 import caliban.{ CalibanError, GraphQLRequest, GraphQLResponse, InputValue }
 import zio._
+import zio.concurrent.{ ConcurrentMap, ConcurrentSet }
 
 import java.nio.charset.StandardCharsets
+import scala.collection.immutable.TreeMap
 import scala.collection.mutable
 
 object ApolloPersistedQueries {
 
   trait ApolloPersistence {
     def get(hash: String): UIO[Option[Document]]
     def add(hash: String, query: Document): UIO[Unit]
+    def isValidated(hash: String, variablesHashCode: Int): UIO[Boolean]
+    def registerValidation(hash: String, variablesHashCode: Int): UIO[Unit]
   }
 
   object ApolloPersistence {
 
-    def get(hash: String): ZIO[ApolloPersistence, Nothing, Option[Document]]      =
+    def get(hash: String): ZIO[ApolloPersistence, Nothing, Option[Document]]                            =
       ZIO.serviceWithZIO[ApolloPersistence](_.get(hash))
-    def add(hash: String, query: Document): ZIO[ApolloPersistence, Nothing, Unit] =
+    def add(hash: String, query: Document): ZIO[ApolloPersistence, Nothing, Unit]                       =
       ZIO.serviceWithZIO[ApolloPersistence](_.add(hash, query))
+    def isValidated(hash: String, variablesHashCode: Int): ZIO[ApolloPersistence, Nothing, Boolean]     =
+      ZIO.serviceWithZIO[ApolloPersistence](_.isValidated(hash, variablesHashCode))
+    def registerValidation(hash: String, variablesHashCode: Int): ZIO[ApolloPersistence, Nothing, Unit] =
+      ZIO.serviceWithZIO[ApolloPersistence](_.registerValidation(hash, variablesHashCode))
 
-    val live: UIO[ApolloPersistence] = Ref.make[Map[String, Document]](Map()).map { cache =>
-      new ApolloPersistence {
-        override def get(hash: String): UIO[Option[Document]]      = cache.get.map(_.get(hash))
-        override def add(hash: String, query: Document): UIO[Unit] = cache.update(_.updated(hash, query))
+    val live: UIO[ApolloPersistence] =
+      (ConcurrentMap.empty[String, Document] <*> ConcurrentSet.empty[(String, Int)]).map {
```

Comment:
I think using `ConcurrentMap[String, Document]` is a better choice here since it avoids copying the underlying Map on each update.

Also, these data structures are ultra thin wrappers around `ConcurrentHashMap`, so we could avoid the `zio-concurrent` dependency and use those directly. Let me know what you think

Reply:
In general I want to limit dependencies but it might be worth it for this one. I wonder if there are any other places this could be used.

Reply:
> I wonder if there are any other places this could be used.

That was my initial thinking as well. I had a look around the codebase but didn't find any obvious ones. The `Ref` that I found that wraps an immutable collection is [this one](https://github.com/ghostdogpr/caliban/blob/cd672ccb6cdb38a015a780f3e7a9da2186bbc196/core/src/main/scala/caliban/execution/Executor.scala#L166), but I think that's fine since it's unlikely to have enough errors to make a difference + prepending on a List doesn't create a whole new object and is very fast.

I took it a step further and had a look at places where we have `ListA :+ element` or `ListA ++ ListB` which are quite bad for performance. There are a couple of places in the codebase although I'm not sure what is the expected size of the lists in those places to justify changing them to a more appropriate collection for append ops like a Vector

Reply:
Yeah I did a round of optimization a while ago and use mutable builders in places it mattered.

Reply:
I think we might need slightly more advanced data structures here that limit the number of entries in both caches. Right now an attacker would be able to DoS the server by sending a lot of unique docs to fill up the caches.

Reply:
> I think we might need slightly more advanced data structures here that limit the number of entries in both caches. Right now an attacker would be able to DoS the server by sending a lot of unique docs to fill up the caches.

That's a good point. In fact at work I provide a custom implementation based on [Caffeine](https://github.com/ben-manes/caffeine) for this reason.

I doubt this wrapper alone justifies bringing in such a dependency, although I think we could detect the number of entries and clear some / all when adding a new key if its size exceeds a certain size. The runtime overhead should be minimal since in most cases new entries should be rare except in the case of an attack.

Reply:
As mentioned on Discord, I changed the implementation to use Java's `ConcurrentHashMap` directly. It felt a bit too much to bring in the `zio.concurrent` dependency for such a thin wrapper